crates/paymaster-execution/src/diagnostics/extractor.rs

```diff
@@ -0,0 +1,199 @@
+    /// These are emitted as OpenTelemetry histograms by the DiagnosticService.
+    #[serde(skip)]
+    pub metrics: Vec<DiagnosticMetric>,
+}
+
+/// A typed value for diagnostic metadata to ensure proper serialization.
+#[derive(Debug, Clone, Serialize)]
+#[serde(untagged)]
+pub enum DiagnosticValue {
+    /// Felt value (serialized as hex string)
+    Felt(Felt),
+    /// String value
+    String(String),
+    /// Numeric value (integer)
+    Number(u128),
+    /// Floating point value (for normalized amounts)
+    Float(f64),
+    /// Boolean value
+    Bool(bool),
+}
+
+impl From<Felt> for DiagnosticValue {
+    fn from(value: Felt) -> Self {
+        Self::Felt(value)
+    }
+}
+
+impl From<String> for DiagnosticValue {
+    fn from(value: String) -> Self {
+        Self::String(value)
+    }
+}
+
+impl From<&str> for DiagnosticValue {
+    fn from(value: &str) -> Self {
+        Self::String(value.to_string())
+    }
+}
+
+impl From<u128> for DiagnosticValue {
+    fn from(value: u128) -> Self {
+        Self::Number(value)
+    }
+}
+
+impl From<bool> for DiagnosticValue {
+    fn from(value: bool) -> Self {
+        Self::Bool(value)
+    }
+}
+
+impl From<f64> for DiagnosticValue {
+    fn from(value: f64) -> Self {
+        Self::Float(value)
+    }
+}
+
+use crate::diagnostics::DiagnosticContext;
+
+/// Trait for extracting metadata from failed transaction calls.
+///
+/// Implementors should focus on a specific contract or protocol and extract
+/// meaningful information when that contract's calls fail.
+///
+/// # Implementation Guidelines
+///
+/// 1. `name()` should return a static string identifying the extractor
+/// 2. `can_handle()` should be fast - check contract addresses and error patterns
+/// 3. `extract()` is only called if `can_handle()` returns true
+/// 4. Use consistent metadata keys for your contract type
+///
+/// # Example
+///
+/// ```ignore
+/// struct MyDexExtractor {
+///     contract_address: Felt,
+/// }
+///
+/// #[async_trait]
+/// impl CallMetadataExtractor for MyDexExtractor {
+///     fn name(&self) -> &'static str { "mydex" }
+///
+///     fn can_handle(&self, ctx: &DiagnosticContext) -> bool {
+///         ctx.calls.iter().any(|c| c.to == self.contract_address)
+///             && ctx.error_message.contains("slippage")
+///     }
+///
+///     async fn extract(&self, ctx: &DiagnosticContext) -> CallDiagnostic {
+///         // Extract swap parameters from calldata...
+///     }
+/// }
+/// ```
+#[async_trait]
+pub trait CallMetadataExtractor: Send + Sync {
+    /// Returns the name of this extractor for logging purposes.
+    fn name(&self) -> &'static str;
+
+    /// Determines if this extractor can handle the given context.
+    ///
+    /// This method should be fast as it's called for each extractor in the registry.
+    /// Check contract addresses, selectors, and/or error message patterns.
+    fn can_handle(&self, context: &DiagnosticContext) -> bool;
```

Comment:
Why do we need that check separate from the "extract" ? Why "extract" cannot return "Option<CallDiagnostic>" and we let the logic completely defined there. It's more flexible in the kind of behavior we can support.